pmlc/dialect/pxa/tests/ops.mlir

```diff
@@ -0,0 +1,15 @@
+// RUN: pmlc-opt -pxa-dataflow-opt -canonicalize -convert-pxa-to-affine %s | FileCheck %s
+
+// CHECK-LABEL: func @pxa_vector_reduce
+func @pxa_vector_reduce(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>, %arg2: memref<100x100xf32>) {
```

Comment:
At the PXA level, outputs are typically returned rather than written into, i.e:

func @pxa_vector_reduce(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>) -> (memref<100x100xf32>)

Then, inside the body, allocate the output (with alloc), add an affine.parallel that clears it via pxa.reduce assign, then do the vector reductions and then return the result of the vector reduction.

Something like

%a = alloc<...>
%b = affine.parallel ... {
  %cst = constant.zero
  %c = pxa.reduce assign %cst, %b[...] ...
  yield %c
}
%d = affine.parallel (existing AP) {
  ...
  %e = pxa.vector_reduce add %vec, %b : ...
  yield %e
}
return %d

Look at the EDSL tests IR for more complete examples, it's a bit weird.  This should lower to very similar code to what you already lower to.

Basically, the current test doesn't actually use the PXA memref-state semantics (and had you added a CSE pass, the vector_reduce would have vanished I think since it's NoSideEffect and it's return is unused)

Reply:
Added more tests. and returning the value now.